File: 30-days-of-code/day-12---inheritance/solution.py
class Person:
    def __init__(self, firstName, lastName, idNumber):
        self.firstName = firstName
        self.lastName = lastName
        self.idNumber = idNumber
    def printPerson(self):
        print("Name:", self.lastName + ",", self.firstName)
        print("ID:", self.idNumber)

class Student(Person):

    def __init__(self, firstName, lastName, idNumber, scores):
        Person.__init__(self, firstName, lastName, idNumber)
        self.testScores = scores

    def calculate(self):
        a = self.average(self.testScores)
        if a >= 90:
            return 'O'
        if a >= 80:
            return 'E'
        if a >= 70:
            return 'A'
        if a >= 55:
            return 'P'
        if a >= 40:
            return 'D'
        return 'T'

    def average(self, arr):
        sum = 0
        for i in arr:
            sum += i
        return sum/len(arr)

line = input().split()
firstName = line[0]
lastName = line[1]
idNum = line[2]
numScores = int(input()) # not needed for Python
scores = list( map(int, input().split()) )
s = Student(firstName, lastName, idNum, scores)
s.printPerson()
print("Grade:", s.calculate())

File: 30-days-of-code/day-12---inheritance/solution.java
import java.util.*;

class Person {
    protected String firstName;
    protected String lastName;
    protected int idNumber;

    // Constructor
    Person(String firstName, String lastName, int identification){
        this.firstName = firstName;
        this.lastName = lastName;
        this.idNumber = identification;
    }

    // Print person data
    public void printPerson(){
        System.out.println(
                "Name: " + lastName + ", " + firstName
                        + 	"\nID: " + idNumber);
    }

}

class Student extends Person{
    private int[] testScores;

    Student(String firstName, String lastName, int identification, int[] scores) {
        super(firstName, lastName, identification);
        this.testScores = scores;
    }

    public char calculate() {
        double a = this.average(this.testScores);
        if (a >= 90) return 'O';
        if (a >= 80) return 'E';
        if (a >= 70) return 'A';
        if (a >= 55) return 'P';
        if (a >= 40) return 'D';
        return 'T';
    }

    public double average(int[] arr) {
        double sum = 0;
        for (int i: arr) {
            sum += i;
        }
        return sum/arr.length;
    }
}

class Solution {
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        String firstName = scan.next();
        String lastName = scan.next();
        int id = scan.nextInt();
        int numScores = scan.nextInt();
        int[] testScores = new int[numScores];
        for(int i = 0; i < numScores; i++){
            testScores[i] = scan.nextInt();
        }
        scan.close();

        Student s = new Student(firstName, lastName, id, testScores);
        s.printPerson();
        System.out.println("Grade: " + s.calculate() );
    }
}
